docs_src/microservices/security/Ch-SSH-Tunneling-HowToSecureDeviceServices.md

```diff
@@ -0,0 +1,185 @@
+# Security for EdgeX Stack
+
+This page describes one of options to secure the EdgeX software stack with running remote device services like device-virtual, device-rest, device-mqtt, and so on, via secure two-way SSH-tunnelings.
+
+## Basic SSH-Tunneling
+
+In this option to secure the EdgeX software stack, SSH tunneling is utilized. The basic idea is to create a secure SSH connection between  a local machine, the primary host, and a remote machine, the secondary, in which some micro-services or applications can be relayed.   In this particular example, the primary host is  running the whole EdgeX core services including core services and security services but **without** any device service.  The device services are running in the secondary or the remote machine.  
+
+The communication is secure because SSH port forwarding connection is encrypted by default.
+
+The SSH communication is established by introducing some extra SSH-related services:
+
+1) device-ssh-proxy: this is the service with ssh client opening up the SSH communication between the primary and the secondary
+
+2) device-ssh-remote: this is actually the SSH server or daemon service together with device services running on the remote machine
+
+The high-level diagram is shown as follows:
+
+![image](ssh-tunneling device.png) "Top level diagram for SSH tunneling for device services"
+
+<TBD> more description needed
+
+## How to- Reference implementation example
+
+### Setup remote running Virtual Machine
+
+In the example setup, `vagrant` is used on the top of `Virtual Box` to set up as the secondary/remote VM. The network port for ssh standard port 22 is mapped into the vm port 2223 and host port also 2223.
+
+Once you have downloaded the vagrant from Hashicorp website, typical vagrant setup for the first time can be done via command `./vagrant init` and it will generate the Vagrant configuration file.
+
+Here is the Vagrant file used to create the remote machine:
+
+[Vagrantfile]: Vagrantfile "remote VM Vagrant file with docker and docker-compose installed"
+
+### SSH Tunneling: Setup the SSH server on the remote machine
+
+Use the example from Docker hub, an example of ssh damenon or ssh server can be setup pretty easily as a Docker container: https://docs.docker.com/engine/examples/running_ssh_service/ 
+
+Note that this one is the ssh server and it is set up using password authentication by default.  In order to authenticate to this ssh server without password prompt, we injected the generated public SSH key from the primary machine via simple login into the ssh server machine first and then created the `authorized_keys` under `~/.ssh` directory.  Without loss of generality, the following command shows how this is accomplished:
```

Comment:
This manual approach isn't scalable.

Reply:
ok, this is a baby step... not the run phase yet

Reply:
i am not sure to pursue the production level of implementation at this time